use half-aggregation for tm messages

akildemir opened the pull request:
for issue #99

kayabaNerve committed:
dalek 4.0

kayabaNerve committed:
cargo update

Moves to a version of Substrate which uses curve25519-dalek 4.0 (not a rc).
Doesn't yet update the repo to curve25519-dalek 4.0 (as a branch does) due
to the official schnorrkel using a conflicting curve25519-dalek. This would
prevent installation of frost-schnorrkel without a patch.

kayabaNerve committed:
Merge branch 'dalek-4.0' into develop

Commit:
use half-aggregation for tm messages

Commit:
fmt

kayabaNerve commented on coordinator/tributary/src/tendermint.rs:
```diff
@@ -41,6 +44,8 @@ use crate::{
   Blockchain, P2p,
 };
 
+const DST: &[u8] = b"Schnorr Aggregator";
```
"Tributary Tendermint Commit Aggregator"?

kayabaNerve commented on coordinator/tributary/src/tendermint.rs:
```diff
@@ -172,8 +176,24 @@ impl SignatureScheme for Validators {
 
-  fn aggregate(sigs: &[Self::Signature]) -> Self::AggregateSignature {
-    sigs.to_vec()
+  // TODO: this function panics if wrong data passed in, should it?
```
I'd make the comment document the fact it panics. It panicking should be fine.

kayabaNerve commented on coordinator/tributary/src/tendermint.rs:
```diff
@@ -183,12 +203,38 @@ impl SignatureScheme for Validators {
+      return false;
+    };
+
+    // get Rs in aggregate
```
We should have `pub fn Rs(&self) ->[C::G]` on Aggregate, sorry.

kayabaNerve commented on coordinator/tributary/tendermint/src/ext.rs:
```diff
@@ -81,7 +81,12 @@ pub trait SignatureScheme: Send + Sync {
+    &self,
+    validators: &[Self::ValidatorId],
+    sigs: &[Self::Signature],
+    msg: &[u8],
```
msg, sigs, would be more standard.

kayabaNerve reviewed (commented):
LGTM other than a few very minor lints :)

Commit:
fix pr comments

kayabaNerve committed:
cargo update

Achieves three notable updates.

1) Resolves RUSTSEC-2022-0093 by updating libp2p-identity.
2) Removes 3 old rand crates via updating ed25519-dalek (a dependency of
libp2p-identity).
3) Sets serde_derive to 1.0.171 via updating to time 0.3.26 which pins at up to
1.0.171.

The last one is the most important. The former two are niceties.

serde_derive, since 1.0.171, ships a non-reproducible binary blob in what's a
complete compromise of supply chain security. This is done in order to reduce
compile times, yet also for the maintainer of serde (dtolnay) to leverage
serde's position as the 8th most downloaded crate to attempt to force changes
to the Rust build pipeline.

While dtolnay's contributions to Rust are respectable, being behind syn, quote,
and proc-macro2 (the top three crates by downloads), along with thiserror,
anyhow, async-trait, and more (I believe also being part of the Rust project),
they have unfortunately decided to refuse to listen to the community on this
issue (or even engage with counter-commentary). Given their political agenda
they seem to try to be accomplishing with force, I'd go as far as to call their
actions terroristic (as they're using the threat of the binary blob as
justification for cargo to ship 'proper' support for binary blobs).

This is arguably representative of dtolnay's past work on watt. watt was a wasm
interpreter to execute a pre-compiled proc macro. This would save the compile
time of proc macros, yet sandbox it so a full binary did not have to be run.

Unfortunately, watt (while decreasing compile times) fails to be a valid
solution to supply chain security (without massive ecosystem changes). It never
implemented reproducible builds for its wasm blobs, and a malicious wasm blob
could still fundamentally compromise a project. The only solution for an end
user to achieve a secure pipeline would be to locally build the project,
verifying the blob aligns, yet doing so would negate all advantages of the
blob.

dtolnay also seems to be giving up their role as a FOSS maintainer given that
serde no longer works in several environments. While FOSS maintainers are not
required to never implement breaking changes, the version number is still 1.0.
While FOSS maintainers are not required to follow semver, releasing a very
notable breaking change *without a new version number* in an ecosystem which
*does follow semver*, then refusing to acknowledge bugs as bugs with their work
does meet my personal definition of "not actively maintaining their existing
work". Maintenance would be to fix bugs, not introduce and ignore.

For now, serde > 1.0.171 has been banned. In the future, we may host a fork
without the blobs (yet with the patches). It may be necessary to ban all of
dtolnay's maintained crates, if they continue to force their agenda as such,
yet I hope this may be resolved within the next week or so.

Sources:

https://github.com/serde-rs/serde/issues/2538 - Binary blob discussion

This includes several reports of various workflows being broken.

https://github.com/serde-rs/serde/issues/2538#issuecomment-1682519944

dtolnay commenting that security should be resolved via Rust toolchain edits,
not via their own work being secure. This is why I say they're trying to
leverage serde in a political game.

https://github.com/serde-rs/serde/issues/2526 - Usage via git broken

dtolnay explicitly asks the submitting user if they'd be willing to advocate
for changes to Rust rather than actually fix the issue they created. This is
further political arm wrestling.

https://github.com/serde-rs/serde/issues/2530 - Usage via Bazel broken

https://github.com/serde-rs/serde/issues/2575 - Unverifiable binary blob

https://github.com/dtolnay/watt - dtolnay's prior work on precompilation

Commit:
add Rs() api to  SchnorrAggregate

kayabaNerve committed:
Correct serai-processor-tests to dalek 4

kayabaNerve committed:
fmt + deny

akildemir committed:
Slash malevolent validators  (#294)

* add slash tx

* ignore unsigned tx replays

* verify that provided evidence is valid

* fix clippy + fmt

* move application tx handling to another module

* partially handle the tendermint txs

* fix pr comments

* support unsigned app txs

* add slash target to the votes

* enforce provided, unsigned, signed tx ordering within a block

* bug fixes

* add unit test for tendermint txs

* bug fixes

* update tests for tendermint txs

* add tx ordering test

* tidy up tx ordering test

* cargo +nightly fmt

* Misc fixes from rebasing

* Finish resolving clippy

* Remove sha3 from tendermint-machine

* Resolve a DoS in SlashEvidence's read

Also moves Evidence from Vec<Message> to (Message, Option<Message>). That
should meet all requirements while being a bit safer.

* Make lazy_static a dev-depend for tributary

* Various small tweaks

One use of sort was inefficient, sorting unsigned || signed when unsigned was
already properly sorted. Given how the unsigned TXs were given a nonce of 0, an
unstable sort may swap places with an unsigned TX and a signed TX with a nonce
of 0 (leading to a faulty block).

The extra protection added here sorts signed, then concats.

* Fix Tributary tests I broke, start review on tendermint/tx.rs

* Finish reviewing everything outside tests and empty_signature

* Remove empty_signature

empty_signature led to corrupted local state histories. Unfortunately, the API
is only sane with a signature.

We now use the actual signature, which risks creating a signature over a
malicious message if we have ever have an invariant producing malicious
messages. Prior, we only signed the message after the local machine confirmed
it was okay per the local view of consensus.

This is tolerated/preferred over a corrupt state history since production of
such messages is already an invariant. TODOs are added to make handling of this
theoretical invariant further robust.

* Remove async_sequential for tokio::test

There was no competition for resources forcing them to be run sequentially.

* Modify block order test to be statistically significant without multiple runs

* Clean tests

---------

Co-authored-by: Luke Parker <lukeparker5132@gmail.com>

kayabaNerve commented:
LGTM. Will rebase and merge.

kayabaNerve committed:
Merge branch 'aggregate-sigs' of github.com:akildemir/serai into aggregate-sigs

kayabaNerve committed:
Add DSTs to Tributary TX sig_hash functions

Prevents conflicts with other systems/other parts of the Tributary.

kayabaNerve commented:
Hmm. The e2e coordinator tests are failing on this branch (not that they're too stable in general). I'm unsure why, yet will need time to debug.

kayabaNerve commented:
Hm. The failures I'm currently looking at may be unrelated to this PR. Will merge and then review.
